Mod2/Lessons/Week4/IntroToORM.md

```diff
@@ -0,0 +1,395 @@
+
+Open the Package Manager Console and type the following command.
+
+```
+add-migration AddRoomTable
+```
+
+### 4. Apply our Migration to update our database
+
+Open the Package Manager Console and type the following command.
+
+```
+update-database
+```
+
+Let's take a look in PGAdmin at what entity framework created for us!
+
+<p align='center'>
+  <img src='../../Images/Week4/one_to_many_from_ef.png'>
+</p>
+
+### Why an ORM is so helpful!
+
+We've already seen some of the benefits of an ORM in action!
+
+Instead of having to write SQL to build our database and tables, we used our ORM to build the tables based on our classes and their relationships.
+
+Tomorrow we'll learn how we can also use the power of an ORM to easily performing CRUD operations on our database (e.g. Add plants, select specific plants, modify plants, delete plants) in our console application!
+
+We'll also learn how to run more advanced queries on the data in our database using LINQ syntax.
+
+<!-- TODO: Would be awesome to fill in a more complicated example of a query to further demonstrate the power of an ORM. -->
+
+## Checks for Understanding
+
+We covered a lot of new concepts today! We'll keep practicing with Entity Framework over the next couple weeks, so totally expected that these terms are still confusing, just give these definitions your best shot!
+
+In your own words, define the following terms.
+
+1. Framework
+1. ORM
+1. Entity Framework
+1. Context
+1. A Migration
+1. Applying a Migration
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+```c#
+    public class MessageLoggerContext : DbContext
+{
+    // Open question, uppercase or lowercase
+    public DbSet<Message> Messages { get; set; }
+    // Pretty sure not needed, but seems good practice to add everything and not rely on cascade
+    public DbSet<User> Users { get; set; }
+    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+        => optionsBuilder.UseNpgsql(@"Host=localhost;Username=postgres;Password=password123;Database=MessageLogger");
+}
+internal class Program
+{
+    static void Main()
+    {
+        using var ctx = new MessageLoggerContext();
+
+        // -- This is a lightweight alternative to creating migrations, This wipes the DB and then creates the tables when they do not exist. Not sure what good practice for teaching is before we get to migrations later in the unit.
+        // ctx.Database.EnsureDeleted();
+        ctx.Database.EnsureCreated();
+        // Looks like this syntax might be better 
+        // new Blog { Url = "http://blogs.msdn.com/adonet" }
+        var myMessage = new Message("does it have a date?");
+        ctx.Messages.Add(myMessage);
+        ctx.SaveChanges();
+    }
+}
+
+public class Message
+	{
+        public int Id { get; set; }
+
+        public string Content { get; set; }
+		// Interesting, if no setter silently isn't creating a database column....
+        public DateTime CreatedAt { get; set; }
+
+	}
+```
```

Comment:
is this sandbox stuff from your own project?

Reply:
Good call, I should delete all of this. It's was old scratch pad stuff.